File: app/modules/downloader/android/src/main/java/com/jointhebodyinstitute/bodyaxis/DownloaderModule.kt
package com.jointhebodyinstitute.bodyaxis

import android.content.Context
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

class DownloaderModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("Downloader")

    Events(
      "downloadProgress",
      "downloadCompleted",
      "downloadFailed",
      "downloadPaused",
      "downloadResumed",
      "downloadCanceled"
    )

    OnCreate {
      DownloaderEvents.attach(this@DownloaderModule)
    }

    AsyncFunction("saveSessionAsync") { payload: Map<String, Any?> ->
      val context = requireContext()
      val db = OfflineDownloadDatabase(context)
      val (session, downloads) = DownloadPayloadParser.parse(context, payload)

      db.upsertSession(session)
      downloads.forEach { db.upsertDownload(it) }

      val queuedIds = downloads
        .mapNotNull { db.getDownload(it.id) }
        .filter { it.status != DownloadStatus.COMPLETED }
        .map { it.id }
      DownloaderService.enqueue(context, queuedIds)

      db.getDownloads(session.id).map { it.toMap() }
    }

    AsyncFunction("getDownloadsAsync") { sessionId: String? ->
      OfflineDownloadDatabase(requireContext())
        .getDownloads(sessionId?.takeIf { it.isNotBlank() })
        .map { it.toMap() }
    }

    AsyncFunction("getSessionsAsync") {
      OfflineDownloadDatabase(requireContext())
        .getSessions()
        .map { it.toMap() }
    }

    AsyncFunction("getSessionAsync") { sessionId: String ->
      OfflineDownloadDatabase(requireContext()).getSession(sessionId)?.toMap()
    }

    AsyncFunction("getDownloadAsync") { downloadId: String ->
      OfflineDownloadDatabase(requireContext()).getDownload(downloadId)?.toMap()
    }

    AsyncFunction("getPlayableUriAsync") { downloadId: String ->
      val download = OfflineDownloadDatabase(requireContext()).getDownload(downloadId)
      if (download?.status == DownloadStatus.COMPLETED && File(download.filePath).exists()) {
        download.toMap()["localUri"] as? String
      } else {
        null
      }
    }

    AsyncFunction("pauseDownloadAsync") { downloadId: String ->
      DownloaderService.pause(requireContext(), downloadId)
      true
    }

    AsyncFunction("resumeDownloadAsync") { downloadId: String ->
      DownloaderService.resume(requireContext(), downloadId)
      true
    }

    AsyncFunction("cancelDownloadAsync") { downloadId: String ->
      DownloaderService.cancel(requireContext(), downloadId)
      true
    }

    AsyncFunction("removeSessionAsync") { sessionId: String ->
      val context = requireContext()
      val removed = OfflineDownloadDatabase(context).removeSession(sessionId)
      removed.forEach {
        File(it.filePath).delete()
        File(it.partialPath).delete()
      }
      true
    }

    AsyncFunction("resumePendingDownloadsAsync") {
      DownloaderService.resumePending(requireContext())
      true
    }
  }

  private fun requireContext(): Context {
    return appContext.reactContext ?: throw IllegalStateException("React context is not available")
  }
}

File: app/modules/downloader/src/DownloaderModule.ts
import { NativeModule, requireNativeModule } from 'expo';

import {
  DownloaderModuleEvents,
  OfflineDownload,
  OfflineSessionRecord,
  SaveOfflineSessionRequest,
} from './Downloader.types';

declare class DownloaderModule extends NativeModule<DownloaderModuleEvents> {
  saveSessionAsync(payload: SaveOfflineSessionRequest): Promise<OfflineDownload[]>;
  getDownloadsAsync(sessionId?: string | null): Promise<OfflineDownload[]>;
  getSessionsAsync(): Promise<OfflineSessionRecord[]>;
  getSessionAsync(sessionId: string): Promise<OfflineSessionRecord | null>;
  getDownloadAsync(downloadId: string): Promise<OfflineDownload | null>;
  getPlayableUriAsync(downloadId: string): Promise<string | null>;
  pauseDownloadAsync(downloadId: string): Promise<boolean>;
  resumeDownloadAsync(downloadId: string): Promise<boolean>;
  cancelDownloadAsync(downloadId: string): Promise<boolean>;
  removeSessionAsync(sessionId: string): Promise<boolean>;
  resumePendingDownloadsAsync(): Promise<boolean>;
}

let moduleInstance: DownloaderModule;
try {
  moduleInstance = requireNativeModule<DownloaderModule>('Downloader');
} catch (error) {
  console.warn("Downloader native module is not available. Using fallback mock.", error);
  moduleInstance = {
    saveSessionAsync: async () => [],
    getDownloadsAsync: async () => [],
    getSessionsAsync: async () => [],
    getSessionAsync: async () => null,
    getDownloadAsync: async () => null,
    getPlayableUriAsync: async () => null,
    pauseDownloadAsync: async () => false,
    resumeDownloadAsync: async () => false,
    cancelDownloadAsync: async () => false,
    removeSessionAsync: async () => false,
    resumePendingDownloadsAsync: async () => false,
    addListener: () => ({ remove: () => {} }),
    removeAllListeners: () => {},
  } as unknown as DownloaderModule;
}

export default moduleInstance;
